@here/harp-text-canvas/lib/typesetting/SmoothPathTypesetter.ts

```diff
@@ -0,0 +1,323 @@
+                if (pathOverflow === true) {
+                    break;
+                } else {
+                    return false;
+                }
+            }
+
+            const tangent = path.getTangent(this.m_tempPathOffset);
+            const angle = Math.atan2(tangent.y, tangent.x);
+
+            const pointsLength = this.m_tempPoints.length;
+            if (pointsLength > 0) {
+                const prevAngle = this.m_tempAngles[pointsLength - 1];
+                this.m_tempAngleDelta.push(MathUtils.circleDistance(angle - prevAngle));
+            }
+
+            this.m_tempPoints.push(textPoint);
+            this.m_tempGlyphs.push(glyphData);
+            this.m_tempAngles.push(angle);
+
+            const isMirrored =
+                UnicodeUtils.isRtlMirrored(glyphData.codePoint) &&
+                direction === UnicodeUtils.Direction.RTL;
+
+            this.m_tempMirrored.push(isMirrored);
+
+            const glyphFont = glyphData.font;
+            const glyphFontMetrics = glyphFont.metrics;
+            const isSmallCaps = this.m_tempSmallCaps
+                ? smallCapsArray![i] && textRenderStyle.fontVariant === FontVariant.SmallCaps
+                : false;
+            const smallCapsScale = isSmallCaps
+                ? glyphFontMetrics.xHeight / glyphFontMetrics.capHeight
+                : 1.0;
+            const glyphScale = this.m_tempScale * smallCapsScale;
+
+            // Advance the current position and proceed to next glyph in the run.
+            this.m_tempPathOffset +=
+                ((glyphData.advanceX + textLayoutStyle.tracking) * glyphScale) /
+                this.m_tempPathLength;
+        }
+
+        return true;
+    }
+
+    private commitLine(): boolean {
+        // Gather common typesetting parameters.
+        const glyphDataArray = this.m_currentParams!.glyphs;
+        const smallCapsArray = this.m_currentParams!.smallCapsArray;
+        const fontCatalog = this.m_currentParams!.fontCatalog;
+        const textRenderStyle = this.m_currentParams!.textRenderStyle;
+        const textLayoutStyle = this.m_currentParams!.textLayoutStyle;
+        const position = this.m_currentParams!.position;
+        const geometry = this.m_currentParams!.geometry;
+        const globalBounds = this.m_currentParams!.globalBounds;
+        const individualBounds = this.m_currentParams!.individualBounds;
+        const vertexBuffer = this.m_currentParams!.vertexBuffer;
+
+        const defaultGlyphRotation = textRenderStyle.rotation;
+        const normalDisplacement =
+            textLayoutStyle.verticalAlignment *
+            glyphDataArray[0].font.metrics.capHeight *
+            this.m_tempScale;
+
+        const e = this.m_tempPoints.length - 1;
+
+        const v1 = new THREE.Vector2();
+        const v2 = new THREE.Vector2();
+        const normal = new THREE.Vector2();
+
+        let currentAngle = this.m_tempAngles[0];
+        for (let i = 0; i < this.m_tempGlyphs.length; i++) {
```

Comment:
Missing some documentation here. Seems like you're doing some low-pass filter with a 1D 5-sample neighborhood, but you should try to add some doc so it's easy to understand and mantain for all devs.

Reply:
Added some notes.